In [1]:
# Imports

import pandas as pd
import os

In [2]:
pwd = os.getcwd()
pwd

'/Users/jasonisberto/Python_Projects/Netflix Movies and TV Shows'

In [3]:
# Increase width of Jupyter Notebook interface

from IPython.core.display import display, HTML
display(HTML("<style>.container { width:100% !important; }</style>"))

In [4]:
netflix_data = pd.read_csv('netflix_titles.csv')
netflix_data

show_id     type                  title         director  \
0         s1    Movie   Dick Johnson Is Dead  Kirsten Johnson   
1         s2  TV Show          Blood & Water              NaN   
2         s3  TV Show              Ganglands  Julien Leclercq   
3         s4  TV Show  Jailbirds New Orleans              NaN   
4         s5  TV Show           Kota Factory              NaN   
...      ...      ...                    ...              ...   
8802   s8803    Movie                 Zodiac    David Fincher   
8803   s8804  TV Show            Zombie Dumb              NaN   
8804   s8805    Movie             Zombieland  Ruben Fleischer   
8805   s8806    Movie                   Zoom     Peter Hewitt   
8806   s8807    Movie                 Zubaan      Mozez Singh   

                                                   cast        country  \
0                                                   NaN  United States   
1     Ama Qamata, Khosi Ngema, Gail Mabalane, Thaban...   South Africa   
2     Sami Bouajila, Tracy Gotoas, Samuel Jouy, Nabi...            NaN   
3                                                   NaN            NaN   
4     Mayur More, Jitendra Kumar, Ranjan Raj, Alam K...          India   
...                                                 ...            ...   
8802  Mark Ruffalo, Jake Gyllenhaal, Robert Downey J...  United States   
8803                                                NaN            NaN   
8804  Jesse Eisenberg, Woody Harrelson, Emma Stone, ...  United States   
8805  Tim Allen, Courteney Cox, Chevy Chase, Kate Ma...  United States   
8806  Vicky Kaushal, Sarah-Jane Dias, Raaghav Chanan...          India   

              date_added  release_year rating   duration  \
0     September 25, 2021          2020  PG-13     90 min   
1     September 24, 2021          2021  TV-MA  2 Seasons   
2     September 24, 2021          2021  TV-MA   1 Season   
3     September 24, 2021          2021  TV-MA   1 Season   
4     September 24, 2021          2021  TV-MA  2 Seasons   
...                  ...           ...    ...        ...   
8802   November 20, 2019          2007      R    158 min   
8803        July 1, 2019          2018  TV-Y7  2 Seasons   
8804    November 1, 2019          2009      R     88 min   
8805    January 11, 2020          2006     PG     88 min   
8806       March 2, 2019          2015  TV-14    111 min   

                                              listed_in  \
0                                         Documentaries   
1       International TV Shows, TV Dramas, TV Mysteries   
2     Crime TV Shows, International TV Shows, TV Act...   
3                                Docuseries, Reality TV   
4     International TV Shows, Romantic TV Shows, TV ...   
...                                                 ...   
8802                     Cult Movies, Dramas, Thrillers   
8803             Kids' TV, Korean TV Shows, TV Comedies   
8804                            Comedies, Horror Movies   
8805                 Children & Family Movies, Comedies   
8806     Dramas, International Movies, Music & Musicals   

                                            description  
0     As her father nears the end of his life, filmm...  
1     After crossing paths at a party, a Cape Town t...  
2     To protect his family from a powerful drug lor...  
3     Feuds, flirtations and toilet talk go down amo...  
4     In a city of coaching centers known to train I...  
...                                                 ...  
8802  A political cartoonist, a crime reporter and a...  
8803  While living alone in a spooky town, a young g...  
8804  Looking to survive in a world taken over by zo...  
8805  Dragged from civilian life, a former superhero...  
8806  A scrappy but poor boy worms his way into a ty...  

[8807 rows x 12 columns]

## Task 1: What is the most popular release year for movies on Netflix?

In [7]:
# My attempt 

netflix_data['release_year'].value_counts()


2018    1147
2017    1032
2019    1030
2020     953
2016     902
        ... 
1966       1
1925       1
1947       1
1959       1
1961       1
Name: release_year, Length: 74, dtype: int64

In [6]:
# Need to filter for just movies, not movies and tv shows

movies_data = netflix_data[netflix_data['type'] == 'Movie']

In [20]:
# Sanity check that this worked

# unique() -- Return unique values based on a hash table

movies_data['type'].unique()

array(['Movie'], dtype=object)

In [21]:
# Sanity check that this worked

movies_data['type'].value_counts()

Movie    6131
Name: type, dtype: int64

In [12]:
movies_data

show_id   type                             title  \
0         s1  Movie              Dick Johnson Is Dead   
6         s7  Movie  My Little Pony: A New Generation   
7         s8  Movie                           Sankofa   
9        s10  Movie                      The Starling   
12       s13  Movie                      Je Suis Karl   
...      ...    ...                               ...   
8801   s8802  Movie                           Zinzana   
8802   s8803  Movie                            Zodiac   
8804   s8805  Movie                        Zombieland   
8805   s8806  Movie                              Zoom   
8806   s8807  Movie                            Zubaan   

                           director  \
0                   Kirsten Johnson   
6     Robert Cullen, José Luis Ucha   
7                      Haile Gerima   
9                    Theodore Melfi   
12              Christian Schwochow   
...                             ...   
8801                Majid Al Ansari   
8802                  David Fincher   
8804                Ruben Fleischer   
8805                   Peter Hewitt   
8806                    Mozez Singh   

                                                   cast  \
0                                                   NaN   
6     Vanessa Hudgens, Kimiko Glenn, James Marsden, ...   
7     Kofi Ghanaba, Oyafunmike Ogunlano, Alexandra D...   
9     Melissa McCarthy, Chris O'Dowd, Kevin Kline, T...   
12    Luna Wedler, Jannis Niewöhner, Milan Peschel, ...   
...                                                 ...   
8801  Ali Suliman, Saleh Bakri, Yasa, Ali Al-Jabri, ...   
8802  Mark Ruffalo, Jake Gyllenhaal, Robert Downey J...   
8804  Jesse Eisenberg, Woody Harrelson, Emma Stone, ...   
8805  Tim Allen, Courteney Cox, Chevy Chase, Kate Ma...   
8806  Vicky Kaushal, Sarah-Jane Dias, Raaghav Chanan...   

                                                country          date_added  \
0                                         United States  September 25, 2021   
6                                                   NaN  September 24, 2021   
7     United States, Ghana, Burkina Faso, United Kin...  September 24, 2021   
9                                         United States  September 24, 2021   
12                              Germany, Czech Republic  September 23, 2021   
...                                                 ...                 ...   
8801                       United Arab Emirates, Jordan       March 9, 2016   
8802                                      United States   November 20, 2019   
8804                                      United States    November 1, 2019   
8805                                      United States    January 11, 2020   
8806                                              India       March 2, 2019   

      release_year rating duration  \
0             2020  PG-13   90 min   
6             2021     PG   91 min   
7             1993  TV-MA  125 min   
9             2021  PG-13  104 min   
12            2021  TV-MA  127 min   
...            ...    ...      ...   
8801          2015  TV-MA   96 min   
8802          2007      R  158 min   
8804          2009      R   88 min   
8805          2006     PG   88 min   
8806          2015  TV-14  111 min   

                                             listed_in  \
0                                        Documentaries   
6                             Children & Family Movies   
7     Dramas, Independent Movies, International Movies   
9                                     Comedies, Dramas   
12                        Dramas, International Movies   
...                                                ...   
8801           Dramas, International Movies, Thrillers   
8802                    Cult Movies, Dramas, Thrillers   
8804                           Comedies, Horror Movies   
8805                Children & Family Movies, Comedies   
8806    Dramas, International Movies, Music & Musicals   

                                       

In [7]:
# Tutorial way with 
# groupby()
# reset_index()

movies_data['count'] = 1

tut_release_years_summed = movies_data.groupby(['release_year']).sum().reset_index()[['release_year', 'count']]

<ipython-input-7-89e5f57ed103>:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  movies_data['count'] = 1


In [42]:
# My way with value_counts()
# Issue - not a df? Can't do a sort_values() since there are no column headers

ji_release_years_summed = movies_data['release_year'].value_counts()
ji_release_years_summed

2018    767
2017    767
2016    658
2019    633
2020    517
       ... 
1947      1
1966      1
1963      1
1959      1
1946      1
Name: release_year, Length: 73, dtype: int64

In [38]:
# sort_values()
tut_release_years_summed.sort_values(['count'], ascending = False)

release_year  count
69          2018    767
68          2017    767
67          2016    658
70          2019    633
71          2020    517
..           ...    ...
12          1961      1
14          1963      1
17          1966      1
5           1947      1
4           1946      1

[73 rows x 2 columns]

#### Task 1 Answer: release_year 2017 or 2018 with 767 movie count each

## Task 2: What year did Netflix release the most content to its platform?

In [44]:
netflix_data.head()

show_id     type                  title         director  \
0      s1    Movie   Dick Johnson Is Dead  Kirsten Johnson   
1      s2  TV Show          Blood & Water              NaN   
2      s3  TV Show              Ganglands  Julien Leclercq   
3      s4  TV Show  Jailbirds New Orleans              NaN   
4      s5  TV Show           Kota Factory              NaN   

                                                cast        country  \
0                                                NaN  United States   
1  Ama Qamata, Khosi Ngema, Gail Mabalane, Thaban...   South Africa   
2  Sami Bouajila, Tracy Gotoas, Samuel Jouy, Nabi...            NaN   
3                                                NaN            NaN   
4  Mayur More, Jitendra Kumar, Ranjan Raj, Alam K...          India   

           date_added  release_year rating   duration  \
0  September 25, 2021          2020  PG-13     90 min   
1  September 24, 2021          2021  TV-MA  2 Seasons   
2  September 24, 2021          2021  TV-MA   1 Season   
3  September 24, 2021          2021  TV-MA   1 Season   
4  September 24, 2021          2021  TV-MA  2 Seasons   

                                           listed_in  \
0                                      Documentaries   
1    International TV Shows, TV Dramas, TV Mysteries   
2  Crime TV Shows, International TV Shows, TV Act...   
3                             Docuseries, Reality TV   
4  International TV Shows, Romantic TV Shows, TV ...   

                                         description  
0  As her father nears the end of his life, filmm...  
1  After crossing paths at a party, a Cape Town t...  
2  To protect his family from a powerful drug lor...  
3  Feuds, flirtations and toilet talk go down amo...  
4  In a city of coaching centers known to train I...

In [45]:
# info()

netflix_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8807 entries, 0 to 8806
Data columns (total 12 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   show_id       8807 non-null   object
 1   type          8807 non-null   object
 2   title         8807 non-null   object
 3   director      6173 non-null   object
 4   cast          7982 non-null   object
 5   country       7976 non-null   object
 6   date_added    8797 non-null   object
 7   release_year  8807 non-null   int64 
 8   rating        8803 non-null   object
 9   duration      8804 non-null   object
 10  listed_in     8807 non-null   object
 11  description   8807 non-null   object
dtypes: int64(1), object(11)
memory usage: 825.8+ KB


In [8]:
# pd.to_datetime() - converting an object column with a date value, to a date format

netflix_data['date_time'] = pd.to_datetime(netflix_data['date_added'])

netflix_data['date_time'].head()

0   2021-09-25
1   2021-09-24
2   2021-09-24
3   2021-09-24
4   2021-09-24
Name: date_time, dtype: datetime64[ns]

In [9]:
netflix_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8807 entries, 0 to 8806
Data columns (total 13 columns):
 #   Column        Non-Null Count  Dtype         
---  ------        --------------  -----         
 0   show_id       8807 non-null   object        
 1   type          8807 non-null   object        
 2   title         8807 non-null   object        
 3   director      6173 non-null   object        
 4   cast          7982 non-null   object        
 5   country       7976 non-null   object        
 6   date_added    8797 non-null   object        
 7   release_year  8807 non-null   int64         
 8   rating        8803 non-null   object        
 9   duration      8804 non-null   object        
 10  listed_in     8807 non-null   object        
 11  description   8807 non-null   object        
 12  date_time     8797 non-null   datetime64[ns]
dtypes: datetime64[ns](1), int64(1), object(11)
memory usage: 894.6+ KB


In [17]:
# dt

netflix_data['date_added_year'] = netflix_data['date_time'].dt.year

In [19]:
netflix_data.head(2)

show_id     type                 title         director  \
0      s1    Movie  Dick Johnson Is Dead  Kirsten Johnson   
1      s2  TV Show         Blood & Water              NaN   

                                                cast        country  \
0                                                NaN  United States   
1  Ama Qamata, Khosi Ngema, Gail Mabalane, Thaban...   South Africa   

           date_added  release_year rating   duration  \
0  September 25, 2021          2020  PG-13     90 min   
1  September 24, 2021          2021  TV-MA  2 Seasons   

                                         listed_in  \
0                                    Documentaries   
1  International TV Shows, TV Dramas, TV Mysteries   

                                         description  date_time  \
0  As her father nears the end of his life, filmm... 2021-09-25   
1  After crossing paths at a party, a Cape Town t... 2021-09-24   

   date_added_year  
0           2021.0  
1           2021.0

In [28]:
netflix_data['count'] = 1

netflix_yearly_content_summed = netflix_data.groupby(['date_added_year']).sum().reset_index()[['date_added_year', 'count']]

In [29]:
netflix_yearly_content_summed.sort_values(['count'], ascending = False)

date_added_year  count
11           2019.0   2016
12           2020.0   1879
10           2018.0   1649
13           2021.0   1498
9            2017.0   1188
8            2016.0    429
7            2015.0     82
6            2014.0     24
3            2011.0     13
5            2013.0     11
4            2012.0      3
0            2008.0      2
1            2009.0      2
2            2010.0      1

#### Task 2 Answer: Year 2016, Netflix released 2016 mediums of content

## Task 3: Independent of year, what is the most popular month to add new content?

In [30]:
netflix_data.head()

show_id     type                  title         director  \
0      s1    Movie   Dick Johnson Is Dead  Kirsten Johnson   
1      s2  TV Show          Blood & Water              NaN   
2      s3  TV Show              Ganglands  Julien Leclercq   
3      s4  TV Show  Jailbirds New Orleans              NaN   
4      s5  TV Show           Kota Factory              NaN   

                                                cast        country  \
0                                                NaN  United States   
1  Ama Qamata, Khosi Ngema, Gail Mabalane, Thaban...   South Africa   
2  Sami Bouajila, Tracy Gotoas, Samuel Jouy, Nabi...            NaN   
3                                                NaN            NaN   
4  Mayur More, Jitendra Kumar, Ranjan Raj, Alam K...          India   

           date_added  release_year rating   duration  \
0  September 25, 2021          2020  PG-13     90 min   
1  September 24, 2021          2021  TV-MA  2 Seasons   
2  September 24, 2021          2021  TV-MA   1 Season   
3  September 24, 2021          2021  TV-MA   1 Season   
4  September 24, 2021          2021  TV-MA  2 Seasons   

                                           listed_in  \
0                                      Documentaries   
1    International TV Shows, TV Dramas, TV Mysteries   
2  Crime TV Shows, International TV Shows, TV Act...   
3                             Docuseries, Reality TV   
4  International TV Shows, Romantic TV Shows, TV ...   

                                         description  date_time  \
0  As her father nears the end of his life, filmm... 2021-09-25   
1  After crossing paths at a party, a Cape Town t... 2021-09-24   
2  To protect his family from a powerful drug lor... 2021-09-24   
3  Feuds, flirtations and toilet talk go down amo... 2021-09-24   
4  In a city of coaching centers known to train I... 2021-09-24   

   date_added_year  count  
0           2021.0      1  
1           2021.0      1  
2           2021.0      1  
3           2021.0      1  
4           2021.0      1

In [32]:
netflix_data['date_added_month'] = netflix_data['date_time'].dt.month

In [33]:
netflix_data.head(2)

show_id     type                 title         director  \
0      s1    Movie  Dick Johnson Is Dead  Kirsten Johnson   
1      s2  TV Show         Blood & Water              NaN   

                                                cast        country  \
0                                                NaN  United States   
1  Ama Qamata, Khosi Ngema, Gail Mabalane, Thaban...   South Africa   

           date_added  release_year rating   duration  \
0  September 25, 2021          2020  PG-13     90 min   
1  September 24, 2021          2021  TV-MA  2 Seasons   

                                         listed_in  \
0                                    Documentaries   
1  International TV Shows, TV Dramas, TV Mysteries   

                                         description  date_time  \
0  As her father nears the end of his life, filmm... 2021-09-25   
1  After crossing paths at a party, a Cape Town t... 2021-09-24   

   date_added_year  count  date_added_month  
0           2021.0      1               9.0  
1           2021.0      1               9.0

In [37]:
# netflix_data['count'] = 1 -- we don't need to redo this again, as it's already been ran in previous cell

netflix_monthly_content_summed = netflix_data.groupby(['date_added_month']).sum().reset_index()[['date_added_month','count']]

In [39]:
netflix_monthly_content_summed.sort_values(['count'], ascending = False)

date_added_month  count
6                7.0    827
11              12.0    813
8                9.0    770
3                4.0    764
9               10.0    760
7                8.0    755
2                3.0    742
0                1.0    738
5                6.0    728
10              11.0    705
4                5.0    632
1                2.0    563

#### Task 3 Answer: July is the month where Netflix releases the most content

## Task 4: What is the movie with the longest title in the data set?

In [40]:
netflix_data.head()

show_id     type                  title         director  \
0      s1    Movie   Dick Johnson Is Dead  Kirsten Johnson   
1      s2  TV Show          Blood & Water              NaN   
2      s3  TV Show              Ganglands  Julien Leclercq   
3      s4  TV Show  Jailbirds New Orleans              NaN   
4      s5  TV Show           Kota Factory              NaN   

                                                cast        country  \
0                                                NaN  United States   
1  Ama Qamata, Khosi Ngema, Gail Mabalane, Thaban...   South Africa   
2  Sami Bouajila, Tracy Gotoas, Samuel Jouy, Nabi...            NaN   
3                                                NaN            NaN   
4  Mayur More, Jitendra Kumar, Ranjan Raj, Alam K...          India   

           date_added  release_year rating   duration  \
0  September 25, 2021          2020  PG-13     90 min   
1  September 24, 2021          2021  TV-MA  2 Seasons   
2  September 24, 2021          2021  TV-MA   1 Season   
3  September 24, 2021          2021  TV-MA   1 Season   
4  September 24, 2021          2021  TV-MA  2 Seasons   

                                           listed_in  \
0                                      Documentaries   
1    International TV Shows, TV Dramas, TV Mysteries   
2  Crime TV Shows, International TV Shows, TV Act...   
3                             Docuseries, Reality TV   
4  International TV Shows, Romantic TV Shows, TV ...   

                                         description  date_time  \
0  As her father nears the end of his life, filmm... 2021-09-25   
1  After crossing paths at a party, a Cape Town t... 2021-09-24   
2  To protect his family from a powerful drug lor... 2021-09-24   
3  Feuds, flirtations and toilet talk go down amo... 2021-09-24   
4  In a city of coaching centers known to train I... 2021-09-24   

   date_added_year  count  date_added_month  
0           2021.0      1               9.0  
1           2021.0      1               9.0  
2           2021.0      1               9.0  
3           2021.0      1               9.0  
4           2021.0      1               9.0

In [42]:
# movies_data = netflix_data[netflix_data['type'] == 'Movie'] -- previous code ran in earlier cell to isolate for only 'Movie' types

movies_data.head()

show_id   type                             title  \
0       s1  Movie              Dick Johnson Is Dead   
6       s7  Movie  My Little Pony: A New Generation   
7       s8  Movie                           Sankofa   
9      s10  Movie                      The Starling   
12     s13  Movie                      Je Suis Karl   

                         director  \
0                 Kirsten Johnson   
6   Robert Cullen, José Luis Ucha   
7                    Haile Gerima   
9                  Theodore Melfi   
12            Christian Schwochow   

                                                 cast  \
0                                                 NaN   
6   Vanessa Hudgens, Kimiko Glenn, James Marsden, ...   
7   Kofi Ghanaba, Oyafunmike Ogunlano, Alexandra D...   
9   Melissa McCarthy, Chris O'Dowd, Kevin Kline, T...   
12  Luna Wedler, Jannis Niewöhner, Milan Peschel, ...   

                                              country          date_added  \
0                                       United States  September 25, 2021   
6                                                 NaN  September 24, 2021   
7   United States, Ghana, Burkina Faso, United Kin...  September 24, 2021   
9                                       United States  September 24, 2021   
12                            Germany, Czech Republic  September 23, 2021   

    release_year rating duration  \
0           2020  PG-13   90 min   
6           2021     PG   91 min   
7           1993  TV-MA  125 min   
9           2021  PG-13  104 min   
12          2021  TV-MA  127 min   

                                           listed_in  \
0                                      Documentaries   
6                           Children & Family Movies   
7   Dramas, Independent Movies, International Movies   
9                                   Comedies, Dramas   
12                      Dramas, International Movies   

                                          description  count  
0   As her father nears the end of his life, filmm...      1  
6   Equestria's divided. But a bright-eyed hero be...      1  
7   On a photo shoot in Ghana, an American model s...      1  
9   A woman adjusting to life after a loss contend...      1  
12  After most of her family is murdered in a terr...      1

In [46]:
# len() -- function is used to compute the length of each element in the Series/Index

movies_data['title_length'] = len(movies_data['title']) # -- this will count the number of rows, which is not what we're looking for 

<ipython-input-46-1aa4643c89a4>:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  movies_data['title_length'] = len(movies_data['title'])


In [47]:
# drop(columns = ['']) -- drop a column from a df

movies_data.drop(columns = ['title_length'])

show_id   type                             title  \
0         s1  Movie              Dick Johnson Is Dead   
6         s7  Movie  My Little Pony: A New Generation   
7         s8  Movie                           Sankofa   
9        s10  Movie                      The Starling   
12       s13  Movie                      Je Suis Karl   
...      ...    ...                               ...   
8801   s8802  Movie                           Zinzana   
8802   s8803  Movie                            Zodiac   
8804   s8805  Movie                        Zombieland   
8805   s8806  Movie                              Zoom   
8806   s8807  Movie                            Zubaan   

                           director  \
0                   Kirsten Johnson   
6     Robert Cullen, José Luis Ucha   
7                      Haile Gerima   
9                    Theodore Melfi   
12              Christian Schwochow   
...                             ...   
8801                Majid Al Ansari   
8802                  David Fincher   
8804                Ruben Fleischer   
8805                   Peter Hewitt   
8806                    Mozez Singh   

                                                   cast  \
0                                                   NaN   
6     Vanessa Hudgens, Kimiko Glenn, James Marsden, ...   
7     Kofi Ghanaba, Oyafunmike Ogunlano, Alexandra D...   
9     Melissa McCarthy, Chris O'Dowd, Kevin Kline, T...   
12    Luna Wedler, Jannis Niewöhner, Milan Peschel, ...   
...                                                 ...   
8801  Ali Suliman, Saleh Bakri, Yasa, Ali Al-Jabri, ...   
8802  Mark Ruffalo, Jake Gyllenhaal, Robert Downey J...   
8804  Jesse Eisenberg, Woody Harrelson, Emma Stone, ...   
8805  Tim Allen, Courteney Cox, Chevy Chase, Kate Ma...   
8806  Vicky Kaushal, Sarah-Jane Dias, Raaghav Chanan...   

                                                country          date_added  \
0                                         United States  September 25, 2021   
6                                                   NaN  September 24, 2021   
7     United States, Ghana, Burkina Faso, United Kin...  September 24, 2021   
9                                         United States  September 24, 2021   
12                              Germany, Czech Republic  September 23, 2021   
...                                                 ...                 ...   
8801                       United Arab Emirates, Jordan       March 9, 2016   
8802                                      United States   November 20, 2019   
8804                                      United States    November 1, 2019   
8805                                      United States    January 11, 2020   
8806                                              India       March 2, 2019   

      release_year rating duration  \
0             2020  PG-13   90 min   
6             2021     PG   91 min   
7             1993  TV-MA  125 min   
9             2021  PG-13  104 min   
12            2021  TV-MA  127 min   
...            ...    ...      ...   
8801          2015  TV-MA   96 min   
8802          2007      R  158 min   
8804          2009      R   88 min   
8805          2006     PG   88 min   
8806          2015  TV-14  111 min   

                                             listed_in  \
0                                        Documentaries   
6                             Children & Family Movies   
7     Dramas, Independent Movies, International Movies   
9                                     Comedies, Dramas   
12                        Dramas, International Movies   
...                                                ...   
8801           Dramas, International Movies, Thrillers   
8802                    Cult Movies, Dramas, Thrillers   
8804                           Comedies, Horror Movies   
8805                Children & Family Movies, Comedies   
8806    Dramas, International Movies, Music & Musicals   

                                       

In [48]:
movies_data.head()

show_id   type                             title  \
0       s1  Movie              Dick Johnson Is Dead   
6       s7  Movie  My Little Pony: A New Generation   
7       s8  Movie                           Sankofa   
9      s10  Movie                      The Starling   
12     s13  Movie                      Je Suis Karl   

                         director  \
0                 Kirsten Johnson   
6   Robert Cullen, José Luis Ucha   
7                    Haile Gerima   
9                  Theodore Melfi   
12            Christian Schwochow   

                                                 cast  \
0                                                 NaN   
6   Vanessa Hudgens, Kimiko Glenn, James Marsden, ...   
7   Kofi Ghanaba, Oyafunmike Ogunlano, Alexandra D...   
9   Melissa McCarthy, Chris O'Dowd, Kevin Kline, T...   
12  Luna Wedler, Jannis Niewöhner, Milan Peschel, ...   

                                              country          date_added  \
0                                       United States  September 25, 2021   
6                                                 NaN  September 24, 2021   
7   United States, Ghana, Burkina Faso, United Kin...  September 24, 2021   
9                                       United States  September 24, 2021   
12                            Germany, Czech Republic  September 23, 2021   

    release_year rating duration  \
0           2020  PG-13   90 min   
6           2021     PG   91 min   
7           1993  TV-MA  125 min   
9           2021  PG-13  104 min   
12          2021  TV-MA  127 min   

                                           listed_in  \
0                                      Documentaries   
6                           Children & Family Movies   
7   Dramas, Independent Movies, International Movies   
9                                   Comedies, Dramas   
12                      Dramas, International Movies   

                                          description  count  title_length  
0   As her father nears the end of his life, filmm...      1          6131  
6   Equestria's divided. But a bright-eyed hero be...      1          6131  
7   On a photo shoot in Ghana, an American model s...      1          6131  
9   A woman adjusting to life after a loss contend...      1          6131  
12  After most of her family is murdered in a terr...      1          6131

In [49]:
movies_data['title_length'] = [len(title) for title in movies_data['title']]

<ipython-input-49-8aa3426f1788>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  movies_data['title_length'] = [len(title) for title in movies_data['title']]


In [50]:
movies_data.head()

show_id   type                             title  \
0       s1  Movie              Dick Johnson Is Dead   
6       s7  Movie  My Little Pony: A New Generation   
7       s8  Movie                           Sankofa   
9      s10  Movie                      The Starling   
12     s13  Movie                      Je Suis Karl   

                         director  \
0                 Kirsten Johnson   
6   Robert Cullen, José Luis Ucha   
7                    Haile Gerima   
9                  Theodore Melfi   
12            Christian Schwochow   

                                                 cast  \
0                                                 NaN   
6   Vanessa Hudgens, Kimiko Glenn, James Marsden, ...   
7   Kofi Ghanaba, Oyafunmike Ogunlano, Alexandra D...   
9   Melissa McCarthy, Chris O'Dowd, Kevin Kline, T...   
12  Luna Wedler, Jannis Niewöhner, Milan Peschel, ...   

                                              country          date_added  \
0                                       United States  September 25, 2021   
6                                                 NaN  September 24, 2021   
7   United States, Ghana, Burkina Faso, United Kin...  September 24, 2021   
9                                       United States  September 24, 2021   
12                            Germany, Czech Republic  September 23, 2021   

    release_year rating duration  \
0           2020  PG-13   90 min   
6           2021     PG   91 min   
7           1993  TV-MA  125 min   
9           2021  PG-13  104 min   
12          2021  TV-MA  127 min   

                                           listed_in  \
0                                      Documentaries   
6                           Children & Family Movies   
7   Dramas, Independent Movies, International Movies   
9                                   Comedies, Dramas   
12                      Dramas, International Movies   

                                          description  count  title_length  
0   As her father nears the end of his life, filmm...      1            20  
6   Equestria's divided. But a bright-eyed hero be...      1            32  
7   On a photo shoot in Ghana, an American model s...      1             7  
9   A woman adjusting to life after a loss contend...      1            12  
12  After most of her family is murdered in a terr...      1            12

In [60]:
top_title_length = movies_data.sort_values(['title_length'], ascending = False)

In [61]:
top_title_length

show_id   type                                              title  \
5164   s5165  Movie  Jim & Andy: The Great Beyond - Featuring a Ver...   
8462   s8463  Movie  The Power of Grayskull: The Definitive History...   
3240   s3241  Movie  Mike Birbiglia: What I Should Have Said Was No...   
4851   s4852  Movie  Steve Martin and Martin Short: An Evening You ...   
3479   s3480  Movie  Cultivating the Seas: History and Future of th...   
...      ...    ...                                                ...   
3668   s3669  Movie                                                 4L   
5959   s5960  Movie                                                 21   
4177   s4178  Movie                                                 IO   
7687   s7688  Movie                                                  P   
5958   s5959  Movie                                                  9   

                           director  \
5164                    Chris Smith   
8462  Randall Lobb, Robert McCallum   
3240                            NaN   
4851                   Marcus Raboy   
3479                    Mana Yasuda   
...                             ...   
3668               Gerardo Olivares   
5959                 Robert Luketic   
4177               Jonathan Helpert   
7687                  Paul Spurrier   
5958                    Shane Acker   

                                                   cast  \
5164                                         Jim Carrey   
8462                                                NaN   
3240                                     Mike Birbiglia   
4851                         Steve Martin, Martin Short   
3479                                                NaN   
...                                                 ...   
3668  Jean Reno, Hovik Keuchkerian, Susana Abaitua, ...   
5959  Jim Sturgess, Kevin Spacey, Kate Bosworth, Aar...   
4177     Margaret Qualley, Anthony Mackie, Danny Huston   
7687  Suangporn Jaturaphut, Opal, Dor Yodrak, Pisama...   
5958  Elijah Wood, John C. Reilly, Jennifer Connelly...   

                       country          date_added  release_year rating  \
5164     Canada, United States   November 17, 2017          2017  TV-MA   
8462     Canada, United States     August 24, 2018          2017  TV-14   
3240             United States   November 22, 2019          2008  TV-MA   
4851             United States        May 25, 2018          2018  TV-14   
3479                       NaN  September 28, 2019          2019  TV-PG   
...                        ...                 ...           ...    ...   
3668                     Spain       July 12, 2019          2019  TV-MA   
5959             United States     January 1, 2020          2008  PG-13   
4177             United States    January 18, 2019          2019  TV-14   
7687  United Kingdom, Thailand        May 31, 2019          2006  TV-MA   
5958             United States   November 16, 2017          2009  PG-13   

     duration                                          listed_in  \
5164   94 min                                      Documentaries   
8462   96 min                                      Documentaries   
3240   56 min                                    Stand-Up Comedy   
4851   74 min                                    Stand-Up Comedy   
3479   45 min                Documentaries, International Movies   
...       ...                                                ...   
3668  105 min               Comedies, Dramas, Independent Movies   
5959  123 min                                             Dramas   
4177   96 min                           Dramas, Sci-Fi & Fantasy   
7687  105 min                Horror Movies, International Movies   
5958   80 min  Action & Adventure, Independent Movies, Sci-Fi...   

                                            description  count  title_length  
5164  Through the lens of his stunningly immersive p...      1           104  
8462  This colorful documentary explores the origins...      1

In [65]:
# iloc -- Purely integer-location based indexing for selection by position.
top_title_length.iloc[0]['title']

'Jim & Andy: The Great Beyond - Featuring a Very Special, Contractually Obligated Mention of Tony Clifton'

#### Task 4 Answer: 'Jim & Andy: The Great Beyond - Featuring a Very Special, Contractually Obligated Mention of Tony Clifton'
title_length: 104

## Task 5: Which actor/actress shows up most frequently in the Netflix data set

In [66]:
netflix_data.head()

show_id     type                  title         director  \
0      s1    Movie   Dick Johnson Is Dead  Kirsten Johnson   
1      s2  TV Show          Blood & Water              NaN   
2      s3  TV Show              Ganglands  Julien Leclercq   
3      s4  TV Show  Jailbirds New Orleans              NaN   
4      s5  TV Show           Kota Factory              NaN   

                                                cast        country  \
0                                                NaN  United States   
1  Ama Qamata, Khosi Ngema, Gail Mabalane, Thaban...   South Africa   
2  Sami Bouajila, Tracy Gotoas, Samuel Jouy, Nabi...            NaN   
3                                                NaN            NaN   
4  Mayur More, Jitendra Kumar, Ranjan Raj, Alam K...          India   

           date_added  release_year rating   duration  \
0  September 25, 2021          2020  PG-13     90 min   
1  September 24, 2021          2021  TV-MA  2 Seasons   
2  September 24, 2021          2021  TV-MA   1 Season   
3  September 24, 2021          2021  TV-MA   1 Season   
4  September 24, 2021          2021  TV-MA  2 Seasons   

                                           listed_in  \
0                                      Documentaries   
1    International TV Shows, TV Dramas, TV Mysteries   
2  Crime TV Shows, International TV Shows, TV Act...   
3                             Docuseries, Reality TV   
4  International TV Shows, Romantic TV Shows, TV ...   

                                         description  date_time  \
0  As her father nears the end of his life, filmm... 2021-09-25   
1  After crossing paths at a party, a Cape Town t... 2021-09-24   
2  To protect his family from a powerful drug lor... 2021-09-24   
3  Feuds, flirtations and toilet talk go down amo... 2021-09-24   
4  In a city of coaching centers known to train I... 2021-09-24   

   date_added_year  count  date_added_month  
0           2021.0      1               9.0  
1           2021.0      1               9.0  
2           2021.0      1               9.0  
3           2021.0      1               9.0  
4           2021.0      1               9.0

In [70]:
# pd.options.display.max_colwidth -- finding out max column width and expanding as needed

pd.options.display.max_colwidth = 150

In [71]:
netflix_data[['title', 'cast']]

title  \
0      Dick Johnson Is Dead   
1             Blood & Water   
2                 Ganglands   
3     Jailbirds New Orleans   
4              Kota Factory   
...                     ...   
8802                 Zodiac   
8803            Zombie Dumb   
8804             Zombieland   
8805                   Zoom   
8806                 Zubaan   

                                                                                                                                                       cast  
0                                                                                                                                                       NaN  
1     Ama Qamata, Khosi Ngema, Gail Mabalane, Thabang Molaba, Dillon Windvogel, Natasha Thahane, Arno Greeff, Xolile Tshabalala, Getmore Sithole, Cindy ...  
2       Sami Bouajila, Tracy Gotoas, Samuel Jouy, Nabiha Akkari, Sofia Lesaffre, Salim Kechiouche, Noureddine Farihi, Geert Van Rampelberg, Bakary Diombera  
3                                                                                                                                                       NaN  
4                                                  Mayur More, Jitendra Kumar, Ranjan Raj, Alam Khan, Ahsaas Channa, Revathi Pillai, Urvi Singh, Arun Kumar  
...                                                                                                                                                     ...  
8802  Mark Ruffalo, Jake Gyllenhaal, Robert Downey Jr., Anthony Edwards, Brian Cox, Elias Koteas, Donal Logue, John Carroll Lynch, Dermot Mulroney, Chlo...  
8803                                                                                                                                                    NaN  
8804                                                    Jesse Eisenberg, Woody Harrelson, Emma Stone, Abigail Breslin, Amber Heard, Bill Murray, Derek Graf  
8805                                Tim Allen, Courteney Cox, Chevy Chase, Kate Mara, Ryan Newman, Michael Cassidy, Spencer Breslin, Rip Torn, Kevin Zegers  
8806                  Vicky Kaushal, Sarah-Jane Dias, Raaghav Chanana, Manish Chaudhary, Meghna Malik, Malkeet Rauni, Anita Shabdish, Chittaranjan Tripathy  

[8807 rows x 2 columns]

In [73]:
# for loop to iterate through the data frame and count things into a dictionary

title_and_cast = netflix_data[['title', 'cast']]

In [83]:
title_and_cast = title_and_cast.dropna()

In [74]:
title_and_cast.head()

title  \
0   Dick Johnson Is Dead   
1          Blood & Water   
2              Ganglands   
3  Jailbirds New Orleans   
4           Kota Factory   

                                                                                                                                                    cast  
0                                                                                                                                                    NaN  
1  Ama Qamata, Khosi Ngema, Gail Mabalane, Thabang Molaba, Dillon Windvogel, Natasha Thahane, Arno Greeff, Xolile Tshabalala, Getmore Sithole, Cindy ...  
2    Sami Bouajila, Tracy Gotoas, Samuel Jouy, Nabiha Akkari, Sofia Lesaffre, Salim Kechiouche, Noureddine Farihi, Geert Van Rampelberg, Bakary Diombera  
3                                                                                                                                                    NaN  
4                                               Mayur More, Jitendra Kumar, Ranjan Raj, Alam Khan, Ahsaas Channa, Revathi Pillai, Urvi Singh, Arun Kumar

In [77]:
# sanity check to make sure your for loop is returning what's expected
for index, row in title_and_cast.iterrows():
    print(index)
    print()
    print(row)
    break

0

title    Dick Johnson Is Dead
cast                      NaN
Name: 0, dtype: object


In [87]:
# itterrows()
# split()
# strip()
# lower()

cast_names_counter = {} # dictionary

for index, row in title_and_cast.iterrows():
    movie_cast = row['cast']
    movie_cast_split = movie_cast.split(',')
    movie_cast_stripped = [name.strip().lower() for name in movie_cast_split]
    print(movie_cast_stripped)
    print()
    if index == 5:
        break

['ama qamata', 'khosi ngema', 'gail mabalane', 'thabang molaba', 'dillon windvogel', 'natasha thahane', 'arno greeff', 'xolile tshabalala', 'getmore sithole', 'cindy mahlangu', 'ryle de morny', 'greteli fincham', 'sello maake ka-ncube', 'odwa gwanya', 'mekaila mathys', 'sandi schultz', 'duane williams', 'shamilla miller', 'patrick mofokeng']

['sami bouajila', 'tracy gotoas', 'samuel jouy', 'nabiha akkari', 'sofia lesaffre', 'salim kechiouche', 'noureddine farihi', 'geert van rampelberg', 'bakary diombera']

['mayur more', 'jitendra kumar', 'ranjan raj', 'alam khan', 'ahsaas channa', 'revathi pillai', 'urvi singh', 'arun kumar']

['kate siegel', 'zach gilford', 'hamish linklater', 'henry thomas', 'kristin lehman', 'samantha sloyan', 'igby rigney', 'rahul kohli', 'annarah cymone', 'annabeth gish', 'alex essoe', 'rahul abburi', 'matt biedel', 'michael trucco', 'crystal balint', 'louis oliver']



In [93]:
# sorted()
# items()
# lambda
# dict()

cast_names_counter = {} # dictionary

for index, row in title_and_cast.iterrows():
    movie_cast = row['cast']
    movie_cast_split = movie_cast.split(',')
    movie_cast_stripped = [name.strip().lower() for name in movie_cast_split]
    for name in movie_cast_stripped:
        cast_names_counter[name] = cast_names_counter.get(name, 0) + 1
        
dict(sorted(cast_names_counter.items(), key = lambda item: item[1], reverse = True))

{'anupam kher': 43,
 'shah rukh khan': 35,
 'julie tejwani': 33,
 'naseeruddin shah': 32,
 'takahiro sakurai': 32,
 'rupa bhimani': 31,
 'akshay kumar': 30,
 'om puri': 30,
 'yuki kaji': 29,
 'amitabh bachchan': 28,
 'paresh rawal': 28,
 'boman irani': 27,
 'rajesh kava': 26,
 'vincent tong': 26,
 'andrea libman': 25,
 'kareena kapoor': 25,
 'samuel l. jackson': 24,
 'john cleese': 24,
 'jigna bhardwaj': 23,
 'fred tatasciore': 23,
 'tara strong': 23,
 'daisuke ono': 22,
 'junichi suwabe': 21,
 'ajay devgn': 21,
 'ashleigh ball': 21,
 'kay kay menon': 21,
 'nicolas cage': 21,
 'nawazuddin siddiqui': 21,
 'adam sandler': 20,
 'david attenborough': 20,
 'salman khan': 20,
 'david spade': 19,
 'yuichi nakamura': 19,
 'erin fitzgerald': 19,
 'mamoru miyano': 19,
 'morgan freeman': 19,
 'james franco': 19,
 'seth rogen': 19,
 'ai kayano': 19,
 'fred armisen': 19,
 'gulshan grover': 19,
 'vatsal dubey': 18,
 'rajesh sharma': 18,
 'fortune feimster': 18,
 'diana kaarina': 18,
 'yashpal sharma

#### Task 5 Answer: Anupam Kher
Appeared in 43 movies